components/central-application-gateway/internal/proxy/proxy.go

```diff
@@ -0,0 +1,208 @@
+	cacheObj, found := p.cache.Get(apiIdentifier.Application, apiIdentifier.Service, apiIdentifier.Entry)
+
+	if found {
+		return cacheObj, nil
+	}
+
+	return p.createCacheEntry(apiIdentifier)
+}
+
+func (p *proxy) createCacheEntry(apiIdentifier model.APIIdentifier) (*CacheEntry, apperrors.AppError) {
+	serviceAPI, err := p.apiExtractor.Get(apiIdentifier)
+	if err != nil {
+		return nil, err
+	}
+
+	proxy, err := makeProxy(serviceAPI.TargetUrl, serviceAPI.RequestParameters, apiIdentifier.Service, p.skipVerify)
+	if err != nil {
+		return nil, err
+	}
+
+	authorizationStrategy := p.newAuthorizationStrategy(serviceAPI.Credentials)
+	csrfTokenStrategy := p.newCSRFTokenStrategy(authorizationStrategy, serviceAPI.Credentials)
+
+	return p.cache.Put(apiIdentifier.Application, apiIdentifier.Service, apiIdentifier.Entry, proxy, authorizationStrategy, csrfTokenStrategy), nil
+}
+
+func (p *proxy) newAuthorizationStrategy(credentials *authorization.Credentials) authorization.Strategy {
+	return p.authorizationStrategyFactory.Create(credentials)
+}
+
+func (p *proxy) newCSRFTokenStrategy(authorizationStrategy authorization.Strategy, credentials *authorization.Credentials) csrf.TokenStrategy {
+	csrfTokenEndpointURL := ""
+	if credentials != nil {
+		csrfTokenEndpointURL = credentials.CSRFTokenEndpointURL
+	}
+	return p.csrfTokenStrategyFactory.Create(authorizationStrategy, csrfTokenEndpointURL)
+}
+
+func (p *proxy) newCSRFTokenStrategyFromCSRFConfig(authorizationStrategy authorization.Strategy, csrfConfig *proxyconfig.CSRFConfig) csrf.TokenStrategy {
+	csrfTokenEndpointURL := ""
+	if csrfConfig != nil {
+		csrfTokenEndpointURL = csrfConfig.TokenURL
+	}
+	return p.csrfTokenStrategyFactory.Create(authorizationStrategy, csrfTokenEndpointURL)
+}
+
+func (p *proxy) setRequestTimeout(r *http.Request) (*http.Request, context.CancelFunc) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(p.proxyTimeout)*time.Second)
+	newRequest := r.WithContext(ctx)
+
+	return newRequest, cancel
+}
+
+func (p *proxy) addAuthorization(r *http.Request, cacheEntry *CacheEntry) apperrors.AppError {
+
+	err := cacheEntry.AuthorizationStrategy.AddAuthorization(r)
+
+	if err != nil {
+		return err
+	}
+
+	return cacheEntry.CSRFTokenStrategy.AddCSRFToken(r)
+}
+
+func (p *proxy) addModifyResponseHandler(r *http.Request, apiIdentifier model.APIIdentifier, cacheEntry *CacheEntry, cacheUpdateFunc updateCacheEntryFunction) apperrors.AppError {
+	// Handle the case when credentials has been changed or OAuth token has expired
+	secondRequestBody, err := copyRequestBody(r)
+	if err != nil {
+		return err
+	}
+
+	modifyResponseFunction := func(response *http.Response) error {
+		retrier := newUnauthorizedResponseRetrier(apiIdentifier, r, secondRequestBody, p.proxyTimeout, cacheUpdateFunc)
+		return retrier.RetryIfFailedToAuthorize(response)
+	}
+
+	cacheEntry.Proxy.ModifyResponse = modifyResponseFunction
```

Comment:
If it is possible, it would be good to have a one place where the proxy parameters are set.
e.g. in the makeProxy method. Otherwise proxy creation is spread through the code and it is not immediatelly visible what it really does.

Reply:
addModifyResponseHandler is invoked on every request, and modifies a shared "cacheEntry.Proxy" object without a mutext

Reply:
As agreed, we will handle it in a separate PR.